tmk_core/common/mousekey.c

```diff
@@ -288,13 +280,158 @@ void mousekey_off(uint8_t code) {
 #endif
 }
 
+#elif defined MK_ACCEL2
 
+#define CLAMP(x, low, high)  (((x) > (high)) ? (high) : (((x) < (low)) ? (low) : (x)))
+#ifndef MOUSEKEY_SLOW_RATIO
+#define MOUSEKEY_SLOW_RATIO 4
+#endif
 
+uint8_t mk_delay = MOUSEKEY_DELAY/10;
+uint8_t mk_interval = MOUSEKEY_INTERVAL;
+uint8_t mk_max_speed = MOUSEKEY_MAX_SPEED;
+uint8_t mk_slow_speed = MOUSEKEY_MAX_SPEED / MOUSEKEY_SLOW_RATIO;
+uint8_t mk_time_to_max = MOUSEKEY_TIME_TO_MAX;
+uint8_t acc_per_int = ((uint16_t) MOUSEKEY_MAX_SPEED) * MOUSEKEY_INTERVAL / MOUSEKEY_TIME_TO_MAX;
+uint8_t mk_wheel_max_speed = MOUSEKEY_WHEEL_MAX_SPEED;
+uint8_t mk_wheel_time_to_max = MOUSEKEY_WHEEL_TIME_TO_MAX;
+
+// x positive right, y positive down
+// +- 1=forward, 2=slow
+int8_t x_state = 0;
+int8_t y_state = 0;
+
+static uint8_t wheel_unit(void)
+{
+    uint16_t unit;
+    if (mousekey_accel & (1<<0)) {
+        unit = (MOUSEKEY_WHEEL_DELTA * mk_wheel_max_speed)/4;
+    } else if (mousekey_accel & (1<<1)) {
+        unit = (MOUSEKEY_WHEEL_DELTA * mk_wheel_max_speed)/2;
+    } else if (mousekey_accel & (1<<2)) {
+        unit = (MOUSEKEY_WHEEL_DELTA * mk_wheel_max_speed);
+    } else if (mousekey_repeat == 0) {
+        unit = MOUSEKEY_WHEEL_DELTA;
+    } else if (mousekey_repeat >= mk_wheel_time_to_max) {
+        unit = MOUSEKEY_WHEEL_DELTA * mk_wheel_max_speed;
+    } else {
+        unit = (MOUSEKEY_WHEEL_DELTA * mk_wheel_max_speed * mousekey_repeat) / mk_wheel_time_to_max;
+    }
+    return (unit > MOUSEKEY_WHEEL_MAX ? MOUSEKEY_WHEEL_MAX : (unit == 0 ? 1 : unit));
+}
+
+static void calc_v(int8_t *state, int8_t *rep)
+{
+    if (!*state) *rep = 0;
+    else if (*state > 0) *rep += acc_per_int;
+    else if (*state < 0) *rep -= acc_per_int;
+    if (*state == 1 || *state == -1) *rep = CLAMP(*rep, -mk_max_speed, mk_max_speed);
+    else *rep = CLAMP(*rep, -mk_slow_speed, mk_slow_speed);
+}
+
+void mousekey_task(void)
+{
+    if (timer_elapsed(last_timer) < (mousekey_repeat ? mk_interval : mk_delay*10))
+        return;
+
+    if (x_state == 0 && y_state == 0 && mouse_report.v == 0 && mouse_report.h == 0)
+        return;
 
-#endif /* #ifndef MK_3_SPEED */
+    if (mousekey_repeat != UINT8_MAX)
+        mousekey_repeat++;
 
+    calc_v(&x_state, &mouse_report.x);
+    calc_v(&y_state, &mouse_report.y);
 
+    if (mouse_report.v > 0) mouse_report.v = wheel_unit();
+    if (mouse_report.v < 0) mouse_report.v = wheel_unit() * -1;
+    if (mouse_report.h > 0) mouse_report.h = wheel_unit();
+    if (mouse_report.h < 0) mouse_report.h = wheel_unit() * -1;
 
+    mousekey_send();
+}
+
+void mousekey_on(uint8_t code)
+{
+    if (code == KC_MS_UP) {
+        if (y_state)  y_state = 2;
+        else          y_state = -1;
+    }
+    else if (code == KC_MS_DOWN) {
+        if (y_state)  y_state = -2;
```

Comment:
can we replace this magic numbers with defines?

Reply:
Do you mean the 2 and -1, I can change that to defines.